src/common/parse.js

```diff
@@ -43,27 +48,55 @@ class ParseHelper {
     return user;
   }
 
-  static signInOrSignUp(appId) {
-    console.log('parse::signInOrSignUp, appId: ', appId);
-    // Set appId as username and password.
-    // No real password is needed because we dont have user authencation in this app. We only want to get access to Parse.User here to access related data
-    const username = appId;
-    const password = appId;
 
-    return Parse.User.logIn(username, password)
-      .catch((err) => {
-        console.log('signInOrSignUp, err: ', err, err.message);
-        if (err.message === 'Invalid username/password.') { // Call sign up if we can't log in using appId
-          console.log(`User not found with appId ${username}. Signing up ...`);
-          const user = new Parse.User();
+  static async signUp() {
+    const newId = await application.createId();
+    const password = randomstring.generate();
+
+    console.log(`parse::signUp, username: ${newId}, password: ${password}`);
+
+    // Sign up
+    const user = new Parse.User();
+    user.set('username', newId);
+    user.set('password', password);
+    await user.signUp();
+
+    // Set user ACL
+    // Protect user information from being read by others
+    user.setACL(new Parse.ACL(user));
+    await user.save();
+
+    // Save username and password to storage
+    await application.saveId(newId);
+    await storage.setUserPassword(password);
+
+    // refresh fcm token
+    fcmHelper.refreshFcmToken();
 
-          user.set('username', username);
-          user.set('password', password);
-          // console.log('DeviceInfo.getDeviceId()', DeviceInfo.getDeviceId());
-          // user.set('deviceId', DeviceInfo.getDeviceId());
+    console.log(`parse::signUp success, username: ${newId}`);
 
-          // TODO: other information needed to be set here.
-          return user.signUp();
+    return user;
+  }
+
+  static async signInOrSignUp(username, password) {
+    console.log(`parse::signInOrSignUp, username: ${username}, password: ${password}`);
+
+    // If password is null, it's new user, call sign up.
+    if (_.isEmpty(password)) {
+      const user = await ParseHelper.signUp();
+      return user;
+    }
+
+    // login
+    return Parse.User.logIn(username, password)
+      .catch(async (err) => {
+        console.log('signInOrSignUp, err: ', JSON.stringify(err), err.message);
+        // If username/password is Invalid, logout and sign up
+        if (err.message === 'Invalid username/password.' || err.message === 'Password must be a string.') { // Call sign up if we can't log in using appId
```

Comment:
This needs to add a check,
"and the username and password were set in the old way". If it's already in the new way, we skip the check otherwise one can repeatedly create.